lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -1712,19 +1713,76 @@ struct RvalueExprVisitor : public ExprVisitor {
     return context.convertAssertionExpression(expr.body, loc);
   }
 
+  // A new class expression can stand for one of two things:
+  // 1) A call to the `new` method (ctor) of a class made outside the scope of
+  // the class
+  // 2) A call to the `super.new` method, i.e. the constructor of the base
+  // class, within the scope of a class, more specifically, within the new
+  // method override of a class.
+  // In the first case we should emit an allocation and a call to the ctor if it
+  // exists (it's optional in System Verilog), in the second case we should emit
+  // a call to the parent's ctor (System Verilog only has single inheritance, so
+  // super is always unambiguous), but no allocation, as the child class' new
+  // invocation already allocated space for both its own and its parent's
+  // properties.
   Value visit(const slang::ast::NewClassExpression &expr) {
     auto type = context.convertType(*expr.type);
-    auto classTy = dyn_cast<moore::ClassHandleType>(type);
-    if (!classTy)
-      return {};
+    circt::moore::ClassHandleType classTy =
+        dyn_cast<moore::ClassHandleType>(type);
+    Value newObj;
+
+    // We are calling new from within a new function, and it's pointing to
+    // super. Check the implicit this ref to figure out the super class type.
+    // Do not allocate a new object.
+    if (!classTy && expr.isSuperClass) {
+      newObj = context.getImplicitThisRef();
+      if (!newObj || !newObj.getType() ||
+          !isa<moore::ClassHandleType>(newObj.getType())) {
+        mlir::emitError(loc) << "Implicit this ref was not set while "
+                                "converting new class function";
```

Comment:
```suggestion
        mlir::emitError(loc) << "implicit this ref was not set while "
                                "converting new class function";
```

Reply:
Maybe this should be an assert, if it's a compiler invariant? Although, making it an error will be easier for debugging 😁

Reply:
We can also do both for good measure 😄